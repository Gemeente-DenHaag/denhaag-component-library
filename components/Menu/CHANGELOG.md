# @gemeente-denhaag/menu

## 0.2.3

### Patch Changes

- Updated dependencies [281fa34]
  - @gemeente-denhaag/icons@1.0.0
  - @gemeente-denhaag/button@0.2.3
  - @gemeente-denhaag/language-switcher@0.2.3
  - @gemeente-denhaag/sheet@0.1.0-alpha.218
