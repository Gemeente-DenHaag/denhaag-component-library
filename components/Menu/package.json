{
  "name": "@gemeente-denhaag/menu",
  "version": "0.2.3-alpha.368",
  "description": "A Menu component",
  "bugs": "https://github.com/nl-design-system/denhaag/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/nl-design-system/denhaag.git",
    "directory": "components/Menu"
  },
  "license": "EUPL-1.2",
  "author": "Municipality of The Hague",
  "exports": {
    ".": {
      "types": "./dist/cjs/index.d.ts",
      "import": "./dist/mjs/index.js",
      "default": "./dist/cjs/index.js"
    },
    "./index.css": "./dist/index.css"
  },
  "main": "dist/cjs/index.js",
  "module": "dist/mjs/index.js",
  "types": "dist/cjs/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c ../../rollup.config.mjs",
    "clean": "rimraf dist tsconfig.tsbuildinfo",
    "watch:components": "chokidar --silent --follow-symlinks \"src/**/*\" -c \"pnpm run build\""
  },
  "dependencies": {
    "@gemeente-denhaag/button": "workspace:*",
    "@gemeente-denhaag/icons": "workspace:*",
    "@gemeente-denhaag/language-switcher": "workspace:*",
    "@gemeente-denhaag/sheet": "workspace:*",
    "@types/uuid": "9.0.7",
    "@utrecht/component-library-css": "3.0.0",
    "uuid": "9.0.1"
  },
  "peerDependencies": {
    "react": "^18.0.0"
  },
  "gitHead": "dcf72a9b79266c1ebede35aff4a02dd9121a980f"
}
