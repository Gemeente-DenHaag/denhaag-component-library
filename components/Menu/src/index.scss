.denhaag-menu-item {
  display: flex;
}
