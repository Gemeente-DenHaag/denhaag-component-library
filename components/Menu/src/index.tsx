import React from 'react';
import './menu.scss';
import BaseProps from '@gemeente-denhaag/baseprops';
import clsx from 'clsx';

export const Menu: React.FC<BaseProps> = (props: Omit<BaseProps, 'classes'>) => {
  const rootClassNames = clsx('denhaag-menu', props.className);
  return (
    <div className={rootClassNames} id={props.id}>
      {props.children}
    </div>
  );
};

export default Menu;
export * from './MenuButton';
