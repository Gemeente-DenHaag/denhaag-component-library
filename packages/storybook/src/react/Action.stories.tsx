import React from 'react';
import { Meta, StoryObj } from '@storybook/react';
import { ActionSingle, ActionSingleProps } from '@gemeente-denhaag/action';
import { FormatDateLabels } from '@gemeente-denhaag/utils';

import readme from '../../../../components/Action/README.md';

const labels: FormatDateLabels = {
  today: 'vandaag',
  yesterday: 'gisteren',
  before: 'vóór',
  approachingDeadline: (daysDifference: number) => {
    if (daysDifference === 1) {
      return `nog ${daysDifference} dag`;
    }
    return `nog ${daysDifference} dagen`;
  },
};

const exampleArgs: ActionSingleProps = {
  children: <strong>Taak</strong>,
  link: '#example',
  labels,
};

const meta = {
  id: 'react-action',
  title: 'React/Actions/Action',
  component: ActionSingle,
  args: exampleArgs,
  tags: ['autodocs'],
  parameters: {
    chromatic: { viewports: [1768, 1280, 768, 360] },
    docs: {
      description: {
        component: readme,
      },
    },
  },
} as Meta<typeof ActionSingle>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};

export const Hover: Story = {
  args: { ...Default.args, className: 'denhaag-action--hover' },
};

export const Focus: Story = {
  args: { ...Default.args, className: 'denhaag-action--focus' },
};

export const WithDate: Story = {
  args: { ...Default.args, dateTime: '2023-09-28T19:47:36.593Z' },
};

export const WithDateRelative: Story = {
  args: { ...Default.args, dateTime: '2023-10-02T19:47:36.593Z', now: '2023-09-28T19:47:36.593Z', relativeDate: true },
};

export const WithDateWarning: Story = {
  args: { ...Default.args, dateTime: '2023-09-30T19:47:36.593Z', now: '2023-09-28T19:47:36.593Z', relativeDate: true },
};

// ### Multiple - hover

// <Canvas>
//   <Story name="Multiple - hover">
//     {() => (
//       <Action
//         actions={
//           <ButtonLink href="#example" className="utrecht-button-link--hover">
//             Actie
//           </ButtonLink>
//         }
//       >
//         <strong>Taak</strong>
//       </Action>
//     )}
//   </Story>
// </Canvas>

// ### Multiple - focus

// <Canvas>
//   <Story name="Multiple - focus">
//     {() => (
//       <Action
//         actions={
//           <ButtonLink href="#example" className="utrecht-button-link--focus">
//             Actie
//           </ButtonLink>
//         }
//       >
//         <strong>Taak</strong>
//       </Action>
//     )}
//   </Story>
// </Canvas>

// ### Multiple - with date

// <Canvas>
//   <Story name="Multiple - with date">
//     {() => (
//       <Action
//         dateTime={"2023-09-28T19:47:36.593Z"}
//         now={"2023-09-28T19:47:36.593Z"}
//         labels={labels}
//         actions={<ButtonLink href="#example">Actie</ButtonLink>}
//       >
//         <strong>Taak</strong>
//       </Action>
//     )}
//   </Story>
// </Canvas>

// ### Multiple - with date - relative

// <Canvas>
//   <Story name="Multiple - with date - relative">
//     {() => (
//       <Action
//         relativeDate
//         dateTime={"2023-10-02T19:47:36.593Z"}
//         now={"2023-09-28T19:47:36.593Z"}
//         labels={labels}
//         actions={<ButtonLink href="#example">Actie</ButtonLink>}
//       >
//         <strong>Taak</strong>
//       </Action>
//     )}
//   </Story>
// </Canvas>

// ### Multiple - with warning date - relative

// <Canvas>
//   <Story name="Multiple - with warning date - relative">
//     {() => (
//       <Action
//         relativeDate
//         dateTime={"2023-09-30T19:47:36.593Z"}
//         now={"2023-09-28T19:47:36.593Z"}
//         labels={labels}
//         actions={<ButtonLink href="#example">Actie</ButtonLink>}
//       >
//         <strong>Taak</strong>
//       </Action>
//     )}
//   </Story>
// </Canvas>
