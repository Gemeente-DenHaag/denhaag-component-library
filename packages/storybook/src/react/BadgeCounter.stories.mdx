import { ArgsTable, Canvas, Description, Meta, Story } from "@storybook/blocks";

import { BadgeCounter } from "@gemeente-denhaag/components-react";
import pkg from "../../../../components/Button/package.json";
import readme from "../../../../components/Button/README.md";

<Meta
  title="React/Data Display/Badge Counter"
  component={BadgeCounter}
  parameters={{
    componentSubtitle: `${pkg.name} - ${pkg.version}`,
    docs: { source: { type: "dynamic" } },
    status: {
      type: "ALPHA",
    },
  }}
/>

<Description markdown={readme} />

## Stories

### Default

<Canvas>
  <Story name="Default">{(args) => <BadgeCounter {...args}>13</BadgeCounter>}</Story>
</Canvas>

<ArgsTable of={BadgeCounter} />
