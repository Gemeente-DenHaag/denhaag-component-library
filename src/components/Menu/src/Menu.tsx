import React from 'react';
import { Menu as MaterialMenu } from '@material-ui/core';
import BaseProps from '@gemeente-denhaag/baseprops';

export interface MenuProps extends BaseProps {
  /**
   * A HTML element, or a function that returns it.
   * It's used to set the position of the menu.
   */
  anchorEl: HTMLElement | null;

  /**
   * If true (Default) will focus the [role="menu"] if no focusable child is found.
   * Disabled children are not focusable.
   * If you set this prop to false focus will be placed on the parent modal container.
   * This has severe accessibility implications and should only be considered if you manage focus otherwise.
   */
  autoFocus?: boolean;

  /**
   * When opening the menu will not focus the active item but the [role="menu"] unless autoFocus is also set to false.
   * Not using the default means not following WAI-ARIA authoring practices.
   * Please be considerate about possible accessibility implications.
   */
  disableAutoFocusItem?: boolean;

  /**
   * Props applied to the MenuList element.
   */
  MenuListProps?: Record<string, unknown>;

  /**
   * Callback fired when the component requests to be closed.
   * @param event The event source of the callback.
   * @param reason Can be "escapeKeyDown", "backdropClick", "tabKeydown".
   */
  onClose?: (event: React.ChangeEvent<unknown>, reason: string) => void;

  /**
   * If true, the menu is visible.
   */
  open: boolean;

  /**
   * classes prop applied to the Popover element.
   */
  PopoverClasses?: Record<string, unknown>;

  /**
   * The length of the transition in ms, or 'auto'
   */
  transitionDuration?: 'auto' | number | { appear?: number; enter?: number; exit?: number };

  /**
   * Props applied to the Transition element.
   */
  TransitionProps?: Record<string, unknown>;

  /**
   * The variant to use.
   */
  variant?: 'menu' | 'selectedMenu';
}

/**
 * Menu displays a list of choices on temporary surfaces.
 * @param props The properties of a Menu component.
 * @constructor Constructs an instance of Menu.
 */
export const Menu: React.FC<MenuProps> = (props: MenuProps) => {
  return <MaterialMenu {...props}>{props.children}</MaterialMenu>;
};

export default Menu;
