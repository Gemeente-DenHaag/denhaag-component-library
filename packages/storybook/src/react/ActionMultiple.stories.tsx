import React from 'react';
import { Meta, StoryObj } from '@storybook/react';
import { ActionMulti } from '@gemeente-denhaag/action';
import { ButtonLink } from '@utrecht/component-library-react';

import readme from '../../../../components/Action/README.md';

const labels = {
  today: 'vandaag',
  yesterday: 'gisteren',
  before: 'vóór',
  approachingDeadline: (daysDifference: number) => {
    if (daysDifference === 1) {
      return `nog ${daysDifference} dag`;
    }
    return `nog ${daysDifference} dagen`;
  },
};

const exampleArgs = {
  labels,
  children: <strong>Taak</strong>,
  actions: <ButtonLink href="#example">Actie</ButtonLink>,
};

const meta = {
  id: 'react-actions-action-multiple',
  title: 'React/Actions/Action Multiple',
  component: ActionMulti,
  args: exampleArgs,
  tags: ['autodocs'],
  parameters: {
    chromatic: { viewports: [1768, 1280, 768, 360] },
    docs: {
      description: {
        component: readme,
      },
    },
  },
} as Meta<typeof ActionMulti>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};

export const Hover: Story = {
  args: {
    ...Default.args,
    actions: (
      <ButtonLink href="#example" className="utrecht-button-link--hover">
        Actie
      </ButtonLink>
    ),
  },
};

export const Focus: Story = {
  args: {
    ...Default.args,
    actions: (
      <ButtonLink href="#example" className="utrecht-button-link--focus">
        Actie
      </ButtonLink>
    ),
  },
};

export const WithDate: Story = {
  args: { ...Default.args, dateTime: '2023-09-28T19:47:36.593Z' },
};

export const WithDateRelative: Story = {
  args: { ...Default.args, dateTime: '2023-10-02T19:47:36.593Z', now: '2023-09-28T19:47:36.593Z', relativeDate: true },
};

export const WithDateWarning: Story = {
  args: { ...Default.args, dateTime: '2023-09-30T19:47:36.593Z', now: '2023-09-28T19:47:36.593Z', relativeDate: true },
};
