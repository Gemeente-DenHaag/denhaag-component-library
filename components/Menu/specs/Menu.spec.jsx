import * as React from 'react';
import Menu, { MenuButton, MenuButtonExpandable } from '../dist';

describe('Menu tests', () => {
  // Starting with functionality tests
  it('can be clicked', () => {
    const props = {
      onClick: () => {
        console.log('Clicked');
      },
    };

    const spy = cy.spy(props, 'onClick').as('Button click');

    cy.mount(<MenuButtonExpandable onClick={props.onClick}>Test button</MenuButtonExpandable>);

    cy.get('.denhaag-menu-button')
      .contains('Test button')
      .click()
      .then(() => expect(spy).to.have.been.calledOnce);

    cy.mount(<MenuButton href="#test">Test button</MenuButton>);
    cy.get('.denhaag-menu-button').contains('Test button').click();
    cy.location('hash').should('eq', '#test');
  });

  it('can be operated with keyboard', () => {
    const props = {
      onClick: () => {
        console.log('Clicked');
      },
    };

    const spy = cy.spy(props, 'onClick').as('Button click');

    cy.mount(<MenuButtonExpandable onClick={props.onClick}>Test button</MenuButtonExpandable>);

    cy.get('.denhaag-menu-button')
      .type('{enter}')
      .then(() => expect(spy).to.have.been.calledOnce);

    cy.mount(<MenuButton href="#test">Test button</MenuButton>);
    cy.get('.denhaag-menu-button').type('{enter}');
    cy.location('hash').should('eq', '#test');
  });

  it('a11y and snapshots', () => {
    cy.snapshots(
      <Menu id="menu">
        <MenuButton href="https://github.com/nl-design-system/denhaag">Label</MenuButton>
        <MenuButton href="https://github.com/nl-design-system/denhaag" active>
          Label 2
        </MenuButton>
        <MenuButtonExpandable>Label 3</MenuButtonExpandable>
        <MenuButtonExpandable active>Label 4</MenuButtonExpandable>
      </Menu>,
      {},
      {
        selector: '.denhaag-menu-button',
        states: {
          hover: 'denhaag-menu-button--hover',
          focus: 'denhaag-menu-button--focus',
        },
      },
    );
  });
});
