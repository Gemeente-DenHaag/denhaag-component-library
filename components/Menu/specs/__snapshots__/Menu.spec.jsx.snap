exports[`Menu tests > a11y and snapshots #0`] = `
<div id="wrapper" style="display: inline-block;">
  <div style="padding: 10px;">
    <div class="denhaag-menu" id="menu">
      <a
        href="https://github.com/nl-design-system/denhaag"
        class="denhaag-menu-button"
        title="Label"
        >Label</a
      ><a
        href="https://github.com/nl-design-system/denhaag"
        class="denhaag-menu-button denhaag-menu-button--active"
        title="Label 2"
        >Label 2</a
      ><button
        class="denhaag-menu-button denhaag-menu-button--expandable"
        title="Label 3"
      >
        Label 3<span class="denhaag-menu-button__chevron"
          ><svg
            width="1em"
            height="1em"
            viewBox="0 0 24 24"
            fill="none"
            xmlns="http://www.w3.org/2000/svg"
            class="denhaag-icon"
            focusable="false"
            aria-hidden="true"
            shape-rendering="auto"
            aria-label="ChevronDownIcon"
          >
            <path
              d="M5.293 8.293a1 1 0 011.414 0L12 13.586l5.293-5.293a1 1 0 111.414 1.414l-6 6a1 1 0 01-1.414 0l-6-6a1 1 0 010-1.414z"
              fill="currentColor"
            ></path></svg
        ></span></button
      ><button
        class="denhaag-menu-button denhaag-menu-button--expandable denhaag-menu-button--active"
        title="Label 4"
      >
        Label 4<span class="denhaag-menu-button__chevron"
          ><svg
            width="1em"
            height="1em"
            viewBox="0 0 24 24"
            fill="none"
            xmlns="http://www.w3.org/2000/svg"
            class="denhaag-icon"
            focusable="false"
            aria-hidden="true"
            shape-rendering="auto"
            aria-label="ChevronDownIcon"
          >
            <path
              d="M5.293 8.293a1 1 0 011.414 0L12 13.586l5.293-5.293a1 1 0 111.414 1.414l-6 6a1 1 0 01-1.414 0l-6-6a1 1 0 010-1.414z"
              fill="currentColor"
            ></path></svg
        ></span>
      </button>
    </div>
  </div>
</div>
`;
